src/Neo/SmartContract/Native/PolicyContract.cs

```diff
@@ -258,6 +269,104 @@ public bool IsBlocked(IReadOnlyStore snapshot, UInt160 account)
             return snapshot.Contains(CreateStorageKey(Prefix_BlockedAccount, account));
         }
 
+        internal bool IsWhitelistFeeContract(DataCache snapshot, UInt160 contractHash, string method, int argCount, [NotNullWhen(true)] out long? fixedFee)
+        {
+            // Check contract existence
+
+            var currentContract = ContractManagement.GetContract(snapshot, contractHash);
+
+            if (currentContract != null)
+            {
+                // Check state existence
+
+                var item = snapshot.TryGet(CreateStorageKey(Prefix_WhitelistedFeeContracts, contractHash, method, argCount));
+
+                if (item != null)
+                {
+                    fixedFee = (long)(BigInteger)item;
+                    return true;
+                }
+            }
+
+            fixedFee = null;
+            return false;
+        }
+
+        /// <summary>
+        /// Remove whitelisted Fee contracts
+        /// </summary>
+        /// <param name="engine">The execution engine.</param>
+        /// <param name="contractHash">The contract to set the whitelist</param>
+        /// <param name="method">Method</param>
+        /// <param name="argCount">Argument count</param>
+        [ContractMethod(Hardfork.HF_Faun, CpuFee = 1 << 15, RequiredCallFlags = CallFlags.States | CallFlags.AllowNotify)]
+        private void RemoveWhitelistFeeContract(ApplicationEngine engine, UInt160 contractHash, string method, int argCount)
+        {
+            if (!CheckCommittee(engine)) throw new InvalidOperationException("Invalid committee signature");
+
+            var key = CreateStorageKey(Prefix_WhitelistedFeeContracts, contractHash, method, argCount);
+
+            if (!engine.SnapshotCache.Contains(key)) throw new InvalidOperationException("Whitelist not found");
+
+            engine.SnapshotCache.Delete(key);
+
+            // Emit event
+            engine.SendNotification(Hash, WhitelistChangedEventName,
+                [new VM.Types.ByteString(contractHash.ToArray()), new VM.Types.ByteString(method.ToStrictUtf8Bytes()),
+                new VM.Types.Integer(argCount), VM.Types.StackItem.Null]);
+        }
+
+        internal int CleanWhitelist(DataCache snapshot, UInt160 contractHash)
+        {
+            var count = 0;
+            var searchKey = CreateStorageKey(Prefix_WhitelistedFeeContracts, contractHash);
+
+            foreach ((var key, _) in snapshot.Find(searchKey, SeekDirection.Forward))
+            {
+                snapshot.Delete(key);
+                count++;
+            }
```

Comment:
Shouldn't we emit notification for every removed entry to match `RemoveWhitelistFeeContract` behaviour?

Reply:
> We can emit the same for add

We already have it. My concern here is `CleanWhitelist` which is called on update.